data-prepper-api/src/main/java/com/amazon/dataprepper/model/event/JacksonEventImpl.java

```diff
@@ -0,0 +1,252 @@
+ *  compatible open source license.
+ *
+ *  Modifications Copyright OpenSearch Contributors. See
+ *  GitHub history for details.
+ */
+
+package com.amazon.dataprepper.model.event;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+
+import static com.fasterxml.jackson.core.JsonPointer.SEPARATOR;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A Jackson Implementation of {@link Event} interface. This implementation relies heavily on JsonNode to manage the keys of the event.
+ * <p>
+ * This implementation supports dot-notation for keys to access nested structures. For example using the key "fizz.buzz" would allow a
+ * user to retrieve the number 42 using {@link #get(String, Class)} from the nested structure below.
+ * <p>
+ *     {
+ *         "foo": "bar"
+ *         "fizz": {
+ *             "buzz": 42
+ *         }
+ *     }
+ *
+ * @since 1.2
+ */
+public class JacksonEventImpl implements Event {
+
+    private static final Logger LOG = LoggerFactory.getLogger(JacksonEventImpl.class);
+
+    private static final String DOT_NOTATION_REGEX = "\\.";
+
+    private final EventMetadata eventMetadata;
+
+    private final JsonNode jsonNode;
+
+    private final ObjectMapper mapper;
+
+    public JacksonEventImpl(final Builder builder) {
+
+        this.mapper = new ObjectMapper();
+
+        if (builder.eventMetadata == null) {
+            this.eventMetadata = new EventMetadataImpl.Builder()
+                    .withEventType(builder.eventType)
+                    .withTimeReceived(builder.timeReceived)
+                    .withAttributes(builder.eventMetadataAttributes)
+                    .build();
+        } else {
+            this.eventMetadata = builder.eventMetadata;
+        }
+
+        if (builder.data == null) {
+            this.jsonNode = mapper.valueToTree(new HashMap<>());
+        } else {
+            this.jsonNode = mapper.valueToTree(builder.data);
+        }
+
+    }
+
+    /**
+     * Adds or updates the key with a given value in the Event.
+     * @param key where the value will be set
+     * @param value value to set the key to
+     * @since 1.2
+     */
+    @Override public void put(final String key, final Object value) {
+
+        checkKeyArgument(key);
+
+        final LinkedList<String> keys = new LinkedList<>(Arrays.asList(key.split(DOT_NOTATION_REGEX)));
+        recursivelyTraverseAndPut(jsonNode, keys, value);
+    }
+
+    private void recursivelyTraverseAndPut(final JsonNode node, final LinkedList<String> keys, final Object value) {
```

Comment:
While safe for right now, you might want to consider the iterative version of traverse so we won't blow the stack.

Reply:
That's a good call. That would require us to enforce a limitation on how many nested structures an event would support.   I would prefer introducing this sooner rather than later. How many iterations should data prepper support?

Reply:
I think right now we should be fine without a limit; but if we want to set one, I'd set it to a depth of `100` or something fairly large that people should not be nesting too.